server/src/main/java/org/candlepin/model/Product.java

```diff
@@ -1407,8 +1407,8 @@ private boolean checkForCycle(Stack<Product> chain, Product child, Collection<Pr
             return true;
         }
 
-        if (children != null && children.stream().anyMatch(product -> product.checkForCycle(chain,
```

Comment:
From the removed hydration code. The queries weren't updated properly for the model changes introduced in CP4, and it had no defensive gating against the scenario where one of the queries didn't find a product for a given UUID, so it would just throw a bunch of nulls into the collection in such a case.

I've added additional filtering to setProvidedProducts to line up with how addProvidedProduct works, as well as an additional test to make sure we don't get into this state again elsewhere.

Reply:
I see, that makes sense! and +1 to the additional filtering in setProviderProducts.